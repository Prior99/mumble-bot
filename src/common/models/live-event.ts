import { is, DataType, oneOf, required, scope, specify } from "hyrest";
import { live } from "../scopes";
import { QueueItem } from "./queue-item";
import { CachedAudio } from "./";

type EventName = "init" |
    "cache add" |
    "cache remove" |
    "queue shift" |
    "queue push" |
    "queue clear";

export class LiveEvent {
    constructor(event: "cache add" | "cache remove", arg2: CachedAudio);
    constructor(event: "queue shift" | "queue push", arg2: QueueItem);
    constructor(event: "queue clear");
    constructor(event: "init", arg2: QueueItem[], arg3: CachedAudio[]);
    constructor(event: EventName, arg2?: QueueItem | CachedAudio | QueueItem[], arg3?: CachedAudio[]) {
        this.event = event;
        switch (event) {
            case "init":
                this.queue = arg2 as QueueItem[];
                this.cachedAudios = arg3;
                break;
            case "cache add":
            case "cache remove":
                this.cachedAudio = arg2 as CachedAudio;
                break;
            case "queue shift":
            case "queue push":
                this.queueItem = arg2 as QueueItem;
                break;
            case "queue clear":
            default:
                break;
        }
    }

    @is(DataType.str).validate(required, oneOf(
        "init",
        "cache add",
        "cache remove",
        "queue shift",
        "queue push",
        "queue clear",
    )) @scope(live)
    public event: EventName;

    @is() @scope(live)
    public queueItem?: QueueItem;

    @is() @scope(live)
    public cachedAudio?: CachedAudio;

    @is() @specify(() => QueueItem) @scope(live)
    public queue?: QueueItem[];

    @is() @specify(() => CachedAudio) @scope(live)
    public cachedAudios?: CachedAudio[];
}
