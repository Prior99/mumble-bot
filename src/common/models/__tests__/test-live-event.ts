import { LiveEvent } from "../live-event";

describe("LiveEvent model", () => {
    describe("the constructor", () => {
        [
            ["cache add", { id: "some-id" } as any],
            ["cache remove", { id: "some-id" } as any],
            ["queue shift", { id: "some-id" } as any],
            ["queue push", { id: "some-id" } as any],
            ["queue clear"],
            [
                "init",
                [
                    { id: "some-id-1" } as any,
                    { id: "some-id-2" } as any,
                ],
                [
                    { id: "some-id-1" } as any,
                    { id: "some-id-2" } as any,
                ],
            ],
            [],
        ].forEach(args => {
            it(`constructs the expected object with arguments "${JSON.stringify(args)}"`, () => {
                expect(new LiveEvent(args[0], args[1], args[3])).toMatchSnapshot();
            });
        });
    });
});
