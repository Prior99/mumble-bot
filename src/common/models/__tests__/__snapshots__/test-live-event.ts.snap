// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveEvent model the constructor constructs the expected object with arguments "["cache add",{"id":"some-id"}]" 1`] = `
LiveEvent {
  "cachedAudio": Object {
    "id": "some-id",
  },
  "event": "cache add",
}
`;

exports[`LiveEvent model the constructor constructs the expected object with arguments "["cache remove",{"id":"some-id"}]" 1`] = `
LiveEvent {
  "cachedAudio": Object {
    "id": "some-id",
  },
  "event": "cache remove",
}
`;

exports[`LiveEvent model the constructor constructs the expected object with arguments "["init",[{"id":"some-id-1"},{"id":"some-id-2"}],[{"id":"some-id-1"},{"id":"some-id-2"}]]" 1`] = `
LiveEvent {
  "cachedAudios": undefined,
  "event": "init",
  "queue": Array [
    Object {
      "id": "some-id-1",
    },
    Object {
      "id": "some-id-2",
    },
  ],
}
`;

exports[`LiveEvent model the constructor constructs the expected object with arguments "["queue clear"]" 1`] = `
LiveEvent {
  "event": "queue clear",
}
`;

exports[`LiveEvent model the constructor constructs the expected object with arguments "["queue push",{"id":"some-id"}]" 1`] = `
LiveEvent {
  "event": "queue push",
  "queueItem": Object {
    "id": "some-id",
  },
}
`;

exports[`LiveEvent model the constructor constructs the expected object with arguments "["queue shift",{"id":"some-id"}]" 1`] = `
LiveEvent {
  "event": "queue shift",
  "queueItem": Object {
    "id": "some-id",
  },
}
`;

exports[`LiveEvent model the constructor constructs the expected object with arguments "[]" 1`] = `LiveEvent {}`;
